### ISSUE TITLE
Running gopls test code lens returns an error

### ISSUE THREAD

**Issue by urandom** [neovim/neovim#14823]
When testing the new code lens support with gopls, I came across the following error when trying to run the "test" code lens:

```
RPC[Error] code_name = unknown, code = 0, message = "DecodeArgs: expected 3 input arguments, got 2 JSON arguments"
```

Other code lenses from gopls seem to work fine.

The lsp server configuration looks like this:
```lua
  local capabilities = lsp_status.capabilities
  capabilities.textDocument.completion.completionItem.snippetSupport = true

  lspconfig['gopls'].setup {
    cmd = {"gopls", "serve"}, -- , "-logfile=/tmp/gopls.log", "-rpc.trace", "--remote=auto"},
    on_attach = on_attach,
    capabilities = capabilities,
    settings = {
      gopls = {
        experimentalPostfixCompletions = true,
        codelenses = {
          generate = true,
          gc_details = true,
          test = true,
          tidy = true,
        },
      },
    },
  }

```

The following minimal code can be used to reproduce the problem:

foo_test.go
```go
package main

import "testing"

func TestHelloWorld(t *testing.T) {
	// t.Fatal("not implemented")
}
```

go.mod
```
module foo

go 1.16
```

The problematic code lens will appear at the `func` line in foo_test.go

On a side note, I noticed that the `LspCodeLens` highlight isn't documented.

**Comment by mfussenegger** [comment]
Looks like this is related to `nil` items being implicitly removed from the command arguments.

This will be fixed by https://github.com/neovim/neovim/pull/14584
With the fix applied, the command definition looks like this:


```
command = {
      arguments = { "file:///path/to/foo_test.go", { "TestHelloWorld" }, vim.NIL },
      command = "gopls.test",
      title = "run test"
    },
```



(Btw does `gopls.test` require a custom client side command? With the fix applied executing it no longer results in an error, but there is also nothing happening. I don't know what's supposed to happen, given that the standard LSP specification doesn't really have an option to show test results)

**Comment by urandom** [comment]
@mfussenegger 
to be honest, i have no idea. I assume that it would run `go test ....` when the lens is ran. if vscode also uses lens for this feature (not really sure), it would the show that command's output

**Comment by ckipp01** [comment]
> Btw does gopls.test require a custom client side command?

I'm going to assume this is the case. I know for Metals we have something similiar with `run`, `test`, `debug` code lenses in order to work required handlers on the client side to basically capture the necessary info to then start a dap session. I wouldn't be surprised at all if this was the same situation here.

**Comment by mfussenegger** [comment]
@ckipp01 Could you also have a look at https://github.com/neovim/neovim/pull/14115 and see if this client command registry would work for the metals code lenses use-cases?

**Comment by alexaandru** [comment]
FWIW, it seems that running tests via code lenses in `gopls` is now deprecated. See https://github.com/golang/tools/blob/master/gopls/doc/settings.md#run-tests-legacy which marks it as "legacy" and see commit https://github.com/golang/tools/commit/8aef11 which introduced that change. 

At the same time, a `gopls.run_tests` command was introduced, so I assume that is meant to be the new way of running tests with `gopls`: https://github.com/golang/tools/blob/master/gopls/doc/commands.md#run-tests

**Comment by mfussenegger** [comment]
Closing this because [the fix](https://github.com/neovim/neovim/pull/14584) got merged a while ago and made it into 0.5.

The tests can be ran with the lenses and there is also output (as a mix of progress report and using `window/showMessage` - so you have to use `:messages` to see it.  

Executing the tests via LSP command work as well with something like this:

```
lua vim.lsp.buf.execute_command({ command = 'gopls.run_tests', arguments = { {URI = vim.uri_from_bufnr(0), Tests = { "TestHelloWorld" } }}})
```

(Making that more convenient would be something for a `nvim-go` LSP extension plugin)
